### DISCUSSION TITLE
[RFC] [Symfony Docs 6.1] PHP Code Examples

### DISCUSSION THREAD

**Discussion by gnito-org** [symfony/symfony#46164] [Q&A]
Would anyone mind if I performed an audit on the PHP code examples in the 6.1 docs and upgraded all those code examples to PHP 8.1?

For example, change all annotations to attributes, use readonly properties, use constructor property promotion, etc., where appropriate.

If okay, should I create a pull request titled [WIP]  so that everyone knows about it?

**Comment by javiereguiluz** [comment]
@gnito-org thanks for this. We update examples from time to time, so I'm OK with this.

But, please let's ask @wouterj, @xabbuh and @OskarStark if they see any problem for doing that in 6.1 branch of the docs. Thanks!

**Reply by wouterj** [reply]
We've always been cautious with introducing new PHP features. We don't want to distract the focus of the code examples away from the Symfony feature they are showing. These are difficult enough for people and using things like constructor property promotion will cause unnecessary confusion (even though they are great features).

So for now, my vote would be -1 to use fancy PHP 8 features in the docs.

**Reply by xabbuh** [reply]
I agree with Wouter.

**Reply by stof** [reply]
However, adding a code example based on attributes everywhere (in a separate tab from the annotations for our multi-format snippets, not as a replacement) makes sense

**Reply by wouterj** [reply]
Indeed. If an attribute tab is missing somewhere, we should add it (in the lowest maintained branch the attribute was introduced to).

**Reply by javiereguiluz** [reply]
Could we detail what we want and don't want for now?

For example:

We want: attributes, constructor property promotion

We don't want: readonly properties, something else?

Thanks!

**Reply by stof** [reply]
Well, from my understanding, Wouter does not want constructor property promotion

**Reply by wouterj** [reply]
My proposal: We don't want to use new PHP features, unless the PHP features are related to a Symfony feature introduced in this version.

E.g. we add PHP attribute examples, as they were added as a Symfony feature. The same applies to Enums: we don't normally use them, unless we want to showcase e.g. the 6.1 Enum-based routing requirement feature.

Other PHP features, which can influence how we write PHP but are not related to Symfony features (e.g. constructor promotion, readonly properties), are not added until the feature is commonly known in the Symfony community. We've never had a real good way to judge whether it was commonly known (and e.g. we've sticked with the `array(...)` PHP syntax for a loooong time). But I would estimate that constructor promotion or readonly properties is not yet common knowledge.

**Reply by gnito-org** [reply]
Symfony 6.1 requires PHP 8.1. That's why I thought it would make sense to upgrade all examples to PHP 8.1.

**Reply by stof** [reply]
Requiring PHP 8.1 is not the same than requiring people reading the doc to be familiar with new features of PHP 8.1

**Reply by gnito-org** [reply]
True, but it could also lead to "oh, is that how you do it in 8.1!" moments for many, especially since they are required to use 8.1 anyway.

There will also be a percentage of the documentation consumers who will wonder why Symfony is not using the features of the PHP version that they require.

Anyway, I'm closing this matter because it seems that the overall consensus is not in favor. Thank you for your feedback.

**Reply by stof** [reply]
> True, but it could also lead to "oh, is that how you do it in 8.1!" moments for many, especially since they are required to use 8.1 anyway.

Being required to use PHP 8.1 does not mean they are required to use those features (or even to know them).
And rather than a "oh, is that how you do it in 8.1!" moment, I fear it would rather be a "what is this weird code?" moment for devs that don't know those features, as they would not understand what they do.

**Reply by OskarStark** [reply]
> Indeed. If an attribute tab is missing somewhere, we should add it (in the lowest maintained branch the attribute was introduced to).

This would be very helpful, but should be contributed to the lowest maintained branch of the feature!

I agree, that we does not need to promote new PHP features unless we document new features or the adoption of the new PHP version is high, but for now (my opinion) it is to early to use all new shiny features, while I am doing so in my codebase.
